dialogue-annotations/src/main/java/com/palantir/dialogue/annotations/MultipartRequestBody.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.annotations;
+
+import com.palantir.dialogue.RequestBody;
+import java.io.IOException;
+import java.io.OutputStream;
+import org.apache.hc.client5.http.entity.mime.AbstractContentBody;
+import org.apache.hc.client5.http.entity.mime.MultipartEntityBuilder;
+import org.apache.hc.client5.http.entity.mime.MultipartPartBuilder;
+import org.apache.hc.core5.http.ContentType;
+import org.apache.hc.core5.http.HttpEntity;
+
+/** An <a href="http://www.ietf.org/rfc/rfc2387.txt">RFC 2387</a>-compliant request body. */
+public final class MultipartRequestBody extends HttpEntityBodyRequestBodyAdapter {
+
+    private MultipartRequestBody(HttpEntity entity) {
+        super(entity);
+    }
+
+    public static final class Builder {
+
+        private final MultipartEntityBuilder builder = MultipartEntityBuilder.create();
+
+        private Builder() {}
+
+        // apache has automatic detection for this, perhaps we shouldn't provide this straight away.
+        // Salt sets it to MIXED even though it only sends form values.
+        public Builder contentType(String contentType) {
+            builder.setContentType(ContentType.create(contentType));
+            return this;
+        }
+
+        public Builder addRequestBodyPart(RequestBodyPartBuilder requestBodyPartBuilder) {
+            builder.addPart(requestBodyPartBuilder.builder.build());
+            return this;
+        }
+
+        public MultipartRequestBody build() {
+            return new MultipartRequestBody(builder.build());
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static RequestBodyPartBuilder requestBodyPartBuilder(RequestBody unsafeRequestBody) {
```

Comment:
Not convinced RequestBody is the right interface here; one client provides InputStreams (roughly), another Files. We should think if supporting Content-Length makes much sense

Reply:
In apache land, the repeatability is inferred by ```#getContentLength() != -1```

Reply:
Fixed